### ISSUE TITLE
A query with complex properties returns data from a wrong table

### ISSUE THREAD

**Issue by oopee** [dotnet/efcore#33161]
### Summary

A query which contains complex properties returns data from a wrong table. This is dangerous as the code runs without errors but returns incorrect data. The issue seems to affect all providers (tested with Sqlite, SQL Server and PostgreSQL). 

### Code

Runnable project: [EF_ComplexProperty_Issue.zip](https://github.com/dotnet/efcore/files/14405601/EF_ComplexProperty_Issue.zip)

Code snippet:
```C#
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.EntityFrameworkCore;

namespace EF_ComplexProperty_Issue;

public static class Program
{
    public static void Main()
    {
        using var ctx = new Context();
        HashSet<int> ids = [1];

        var query = ctx.C
            .Where(x => ids.Contains(x.B.AId.Value))
            .Take(10)
            .Select(x => new
            {
                x.B.A.Id,
                x.B.Info.Created,
            });

        string queryString = query.ToQueryString();
        Console.WriteLine(queryString);
        // The query returns C.Info_Created instead of B.Info_Created for expression 'x.B.Info.Created'.
    }
}

public class Context : DbContext
{
    public DbSet<A> A { get; set; }
    public DbSet<B> B { get; set; }
    public DbSet<C> C { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder options)
        => options.UseSqlite();

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<B>(x => x.ComplexProperty(b => b.Info).IsRequired());
        modelBuilder.Entity<C>(x => x.ComplexProperty(c => c.Info).IsRequired());
    }
}

public class Metadata
{
    public DateTime Created { get; set; }
}

public class A
{
    public int Id { get; set; }
}

public class B
{
    public int Id { get; set; }
    public Metadata Info { get; set; }
    public int? AId { get; set; }

    public A A { get; set; }
}

public class C
{
    public int Id { get; set; }
    public Metadata Info { get; set; }
    public int BId { get; set; }

    public B B { get; set; }
}
```

### Actual output

The inner query incorrectly selects **c.Info_Created** instead of **b.Info_Created**.

```SQL
SELECT "a"."Id", "t"."Info_Created" AS "Created"
FROM (
    SELECT "c"."Info_Created", "b"."AId"
    FROM "C" AS "c"
    INNER JOIN "B" AS "b" ON "c"."BId" = "b"."Id"
    WHERE "b"."AId" = 1
    LIMIT @__p_1
) AS "t"
LEFT JOIN "A" AS "a" ON "t"."AId" = "a"."Id"
```

### Expected output

```SQL
SELECT "a"."Id", "t"."Info_Created" AS "Created"
FROM (
    SELECT "b"."Info_Created", "b"."AId"
    FROM "C" AS "c"
    INNER JOIN "B" AS "b" ON "c"."BId" = "b"."Id"
    WHERE "b"."AId" = 1
    LIMIT @__p_1
) AS "t"
LEFT JOIN "A" AS "a" ON "t"."AId" = "a"."Id"
```

### Provider and version information

EF Core version: 8.0.2
Database provider: Microsoft.EntityFrameworkCore.SqlServer, Microsoft.EntityFrameworkCore.Sqlite, Npgsql.EntityFrameworkCore.PostgreSQL
Target framework: .NET 8.0
Operating system: Windows 10
IDE: Visual Studio 2022 17.8

**Comment by roji** [comment]
Looks like a dup of #32911, @maumar what do you think? We should consider backporting the fix if so.

**Comment by oopee** [comment]
Please consider backporting the fix as this bug is a showstopper for our large, long-standing LOB application. This is the last issue related to our EF6 -> EF Core port.

**Comment by maumar** [comment]
this is indeed a dupe of #32911 I verified that it works on current main bits. Closing

**Comment by maumar** [comment]
@oopee the bug has been approved for patching - fix will be available in 8.0.4 release

**Comment by oopee** [comment]
> @oopee the bug has been approved for patching - fix will be available in 8.0.4 release

Excellent news, thank you!
